projects/client-ws/connector.ts

```diff
@@ -0,0 +1,66 @@
+import { ClientConnector, connectClient, applyApi, Client } from '@remixproject/plugin/connector'
+import { PluginClient } from '@remixproject/plugin/client'
+import { Message } from '../utils/src/types/message'
+import { RemixApi } from '../utils/src/api/remix-profile'
+import { Api, ApiMap } from '../utils/src/types/api'
+
+
+export interface WS {
+  send(data: string): void
+  on(type: 'message', cb: (event: string) => any): this
+}
+
+/**
+ * This Websocket connector works with the library `ws`
+ */
+export class WebsocketConnector implements ClientConnector {
+
+  constructor(private websocket: WS) {}
+
+  /** Send a message to the engine */
+  send(message: Partial<Message>) {
+    this.websocket.send(JSON.stringify(message))
+  }
+
+  /** Get messae from the engine */
+  on(cb: (message: Partial<Message>) => void) {
+    this.websocket.on('message', (event) => cb(JSON.parse(event)))
+  }
+}
+
+/**
+ * Connect an Iframe client to a web engine
+ * @param client An optional iframe client to connect to the engine.
+ *
+ * ---------
+ * @example
+ * ```typescript
+ * const wss = new WebSocket.Server({ port: 8080 });
+ * wss.on('connection', (ws) => {
+ *  const client = createWebsocketClient(ws)
+ * })
+ * ```
+ * ---------
+ * @example
+ * ```typescript
+ * class MyPlugin extends PluginClient {
+ *  methods = ['hello']
+ *  hello() {
+ *   console.log('Hello World')
+ *  }
+ * }
+ * const wss = new WebSocket.Server({ port: 8080 });
```

Comment:
Where does the WebSocket connection `new WebSocket('ws://localhost:8080', 'echo-protocol')` (the engine trying to connect to the websocket client) actually happen?

Reply:
In the server. It's up to the developer to create a websocket connection. I could do it, but I would need to know the client address, and the websocket library. So in the end I think it's better to let the user create his connection and pass it to the plugin client. I'm open to better solutions :)

Reply:
> In the server

websocket server? not sure how that'd work because the connection is only initiated by remix IDE when the websocket plugin is activated.
The websocket address where to connect to is then given by the `url` property of the profile.

Reply:
Ah sorry, I missunderstood your question. What we call "client" in remix-plugin is the plugin side (which in this case is a websocket server).
The code running inside RemixIDE is the Engine, where we create a new websocket connection when the plugin is activated.